rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+flexible, however, as many rows may not be updated in a long time (though most of those rows will be
+in the base layer of the LSM tree anyway, so it may not be so bad).
+
+### SlateDB::Get
+
+SlateDB currently returns a `Option<Bytes>` on a successful read operation. Now that additional
+metadata (timestamps in this RFC) are introduced, we will need to expand this API to include
+more information in the return value. We have a few options here (TODO: update this section when
+we converge on the approach we want):
+
+1. We can return a strongly typed result (such as `Option<Row>` that is a struct with the result as
+   well as any additional retrieved metadata)
+2. We can introduce a `db#get_with_metadata` that will return a tuple of `Bytes` and corresponding
+   metadata
+3. Do nothing (make the timestamp opaque to the user and require them to interact with timestamps
+   only via the public API)
+
+## Implementation
+
+### Data Format
+
+Currently, SlateDB does not have a mechanism for backwards-compatibility of the SST data format.
+We will first introduce the following two fields to the flatbuffer definitions: 
+- `format_version`: field that specifies the codec version that encoded this SST
+- `meta_features`: an enumeration of metadata fields that are available with each encoded row. This
+  makes it possible in the future to save space per row by disabling metadata features, but that is
+  out of scope for this RFC
+
+```fbs
+enum MetaFeatures: byte {
+    Timestamp,
+    TimeToLive,
+}
+
+table SsTableInfo {
+    // ...
+    
+    // the current encoding version of the data block
+    format_version: uint;
+    
+    // the metadata features that are encoded with each row, in order that they are encoded
+    meta_features: [MetaFeatures];
+}
+```
+
+This value will be written when the SST is created. We will read this back using a codec that maps
+1:1 with the format version. The version 0 has the following per-value format:
+
+```
+|  u16    | var |    u32    |  var  |
+|---------|-----|-----------|-------|
+| key_len | key | value_len | value |
+|---------|-----|-----------|-------|
+```
+
+The `format_version` will be bumped to 1 and will be modified to have the following schema:
+
+```
+|  u16    | var | var  |    u32    |  var  |
+|---------|-----|------|-----------|-------|
+| key_len | key | meta | value_len | value |
+|---------|-----|------|-----------|-------|
+```
+
+The newly introduce `meta` field will be decoded using the `meta_features` array specified for
+this SST. In cases where both `Timestamp` and `TimeToLive` are enabled, the row will look like:
+
+```
+|  u16    | var | i64 | i64 |    u32    |  var  |
+|---------|-----|-----|-----|-----------|-------|
+| key_len | key | ts  | ttl | value_len | value |
+|---------|-----|-----|-----|-----------|-------|
+```
+
+Both `ts` and `ttl` are encoded as `i64` in milliseconds. The expiry time is `ts + ttl`.
```

Comment:
Considering we are introducing MVCC for slateDB, if I understand correctly, `ts` will be the system clock? I don't think it's a good idea.

Reply:
This isn't intended for MVCC at the moment but is scoped to support TTL, MVCC using timestamps is slated for a future follow-up. 

With MVCC we need to have an option to keep multiple copies of a key, and then specify at retrieval time the condition with which to keep/ignore a key and during compaction when to remove an old version of a key.

This is a stepping stone for that, but because of the problem of out-of-order inserts we considered it out of scope. I don't imagine anything in this RFC prevents proper MVCC going forward.

Reply:
So you mean we would use the field `timestamp(ts)` for TTL and `seqnum` for MVCC? Maybe it's a matter of habit, I'm used to using `timestamp` for MVCC most of the time. So I may have misunderstood here.

Reply:
yup! I'm taking a page from RocksDB terminology: https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp

You could use the user defined timestamp as a mechanism for implementing userspace MVCC (which is what RocksDB uses it for) but then you require a clock that is monotonically increasing (which is what this RFC suggests). 

Within RocksDB, they use sequence numbers for consistency.